src/app-layer-smb-tcp-rust.c

```diff
@@ -0,0 +1,276 @@
+{
+    SCLogDebug("RustSMBTCPProbe");
+/*
+    if (len == 0 || len < sizeof(SMBHeader)) {
+        return ALPROTO_UNKNOWN;
+    }
+*/
+    // Validate and return ALPROTO_FAILED if needed.
+    if (!rs_smb_probe_tcp_ts(input, len)) {
+        return ALPROTO_FAILED;
+    }
+
+    return ALPROTO_SMB;
+}
+
+static uint16_t RustSMBTCPProbeTC(Flow *f,
+        uint8_t *input, uint32_t len, uint32_t *offset)
+{
+    SCLogDebug("RustSMBTCPProbe");
+/*
+    if (len == 0 || len < sizeof(SMBHeader)) {
+        return ALPROTO_UNKNOWN;
+    }
+*/
+    // Validate and return ALPROTO_FAILED if needed.
+    if (!rs_smb_probe_tcp_tc(input, len)) {
+        return ALPROTO_FAILED;
+    }
+
+    return ALPROTO_SMB;
+}
+
+static int RustSMBGetAlstateProgress(void *tx, uint8_t direction)
+{
+    return rs_smb_tx_get_alstate_progress(tx, direction);
+}
+
+static uint64_t RustSMBGetTxCnt(void *alstate)
+{
+    return rs_smb_state_get_tx_count(alstate);
+}
+
+static void *RustSMBGetTx(void *alstate, uint64_t tx_id)
+{
+    return rs_smb_state_get_tx(alstate, tx_id);
+}
+
+static AppLayerGetTxIterTuple RustSMBGetTxIterator(
+        const uint8_t ipproto, const AppProto alproto,
+        void *alstate, uint64_t min_tx_id, uint64_t max_tx_id,
+        AppLayerGetTxIterState *istate)
+{
+    return rs_smb_state_get_tx_iterator(alstate, min_tx_id, (uint64_t *)istate);
+}
+
+
+static void RustSMBSetTxLogged(void *alstate, void *tx, uint32_t logger)
+{
+    rs_smb_tx_set_logged(alstate, tx, logger);
+}
+
+static LoggerId RustSMBGetTxLogged(void *alstate, void *tx)
+{
+    return rs_smb_tx_get_logged(alstate, tx);
+}
+
+static void RustSMBStateTransactionFree(void *state, uint64_t tx_id)
+{
+    rs_smb_state_tx_free(state, tx_id);
+}
+
+static DetectEngineState *RustSMBGetTxDetectState(void *tx)
+{
+    return rs_smb_state_get_tx_detect_state(tx);
+}
+
+static int RustSMBSetTxDetectState(void *tx, DetectEngineState *s)
+{
+    rs_smb_state_set_tx_detect_state(tx, s);
+    return 0;
+}
+
+static FileContainer *RustSMBGetFiles(void *state, uint8_t direction)
+{
+    return rs_smb_getfiles(direction, state);
+}
+
+static AppLayerDecoderEvents *RustSMBGetEvents(void *state, uint64_t id)
+{
+    return rs_smb_state_get_events(state, id);
+}
+
+static int RustSMBGetEventInfo(const char *event_name, int *event_id,
+    AppLayerEventType *event_type)
+{
+    return rs_smb_state_get_event_info(event_name, event_id, event_type);
+}
+
+static void RustSMBSetDetectFlags(void *tx, uint8_t dir, uint64_t flags)
+{
+    rs_smb_tx_set_detect_flags(tx, dir, flags);
+}
+
+static uint64_t RustSMBGetDetectFlags(void *tx, uint8_t dir)
+{
+    return rs_smb_tx_get_detect_flags(tx, dir);
+}
+
+static void RustSMBStateTruncate(void *state, uint8_t direction)
+{
+    return rs_smb_state_truncate(state, direction);
+}
+
+static StreamingBufferConfig sbcfg = STREAMING_BUFFER_CONFIG_INITIALIZER;
+static SuricataFileContext sfc = { &sbcfg };
+
+void RegisterRustSMBTCPParsers(void)
```

Comment:
This could be simplified using functions from `app-layer-register.h` ?

Reply:
I agree it would be nice but I consider it 'nice to have' for now. Want to get it complete enough feature wise first :)